File: server/src/utils/parse.ts
// import {
//   DocumentsInterface,
//   DocumentsObjectInterface
// } from '@/services/vector-store';

// export const processDocumentsIntoObjects = (
//   documents: DocumentsInterface
// ): DocumentsObjectInterface[] => {
//   return documents.documents
//     .map((document, i) => {
//       return {
//         metadata: documents.metadatas ? documents.metadatas[i] : {},
//         embedding: documents.embeddings ? documents.embeddings[i] : {},
//         document,
//         id: documents.ids[i]
//       };
//     })
//     .sort((a, b) => a.id.localeCompare(b.id, undefined, { numeric: true }));
// };

import {
  DocumentsInterface,
  DocumentsObjectInterface
} from '@/services/vector-store';
import { GetResponse } from 'chromadb/dist/main/types';

export const processDocumentsIntoObjects = (
  documents: GetResponse
): Record<string, DocumentsObjectInterface[]> => {
  const objects: Record<string, DocumentsObjectInterface[]> = {};

  documents.documents.forEach((document, i) => {
    const metadata = documents.metadatas[i] || undefined;
    const embedding = documents.embeddings ? documents.embeddings[i] : {};
    const id = documents.ids[i];
    const filename = metadata?.filename as string;
    const documentString = document as string;
    if (!objects[filename]) {
      objects[filename] = [];
    }

    objects[filename].push({
      metadata,
      embedding,
      document: documentString,
      id
    });
  });

  return objects;
};

export default processDocumentsIntoObjects;

File: pyServer/server/utils/parse.py
from chromadb.api.types import (
    Embedding,
    Metadata,
    Metadata,
    Document,
    GetResult,
)
from typing import Optional, Dict


class DocumentsObjectInterface:
    def __init__(
        self,
        id: str,
        document: Optional[Document] = None,
        metadata: Optional[Metadata] = None,
        embedding: Optional[Embedding] = None,
    ):
        self.metadata = metadata or None
        self.embedding = embedding or None
        self.document = document or None
        self.id = id

    def to_dict(self):
        return {
            "metadata": self.metadata,
            "embedding": self.embedding,
            "document": self.document,
            "id": self.id,
        }


class ProcessedDocumentReturnObject:
    id: str
    document: Optional[Document]
    metadata: Optional[Metadata]
    embedding: Optional[Embedding]


def process_documents_into_objects(
    documents: GetResult,
) -> Dict[str, list[ProcessedDocumentReturnObject]]:
    objects = {}
    docs = documents.get("documents", [])
    metadatas = documents.get("metadatas", [])
    embeddings = documents.get("embeddings", [])
    ids = documents.get("ids", [])

    if not docs:
        return {}

    for i, document in enumerate(docs):
        metadata = metadatas[i] if metadatas else None
        embedding = embeddings[i] if embeddings else None
        doc_id = ids[i]
        filename = (
            str(metadata.get("filename"))
            if metadata and "filename" in metadata
            else "unknown"
        )

        if filename not in objects:
            objects[str(filename)] = []
        else:
            objects[filename].append(
                DocumentsObjectInterface(
                    id=doc_id, document=document, metadata=metadata, embedding=embedding
                ).to_dict()
            )

    return objects
